src/Nethermind/Nethermind.Consensus/Processing/CensorshipDetector.cs

```diff
@@ -0,0 +1,108 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Nethermind.Blockchain;
+using Nethermind.Core;
+using Nethermind.Core.Caching;
+using Nethermind.Core.Collections;
+using Nethermind.Int256;
+using Nethermind.Logging;
+using Nethermind.TxPool;
+
+namespace Nethermind.Consensus.Processing;
+
+public class CensorshipDetector : IDisposable
+{
+    private readonly ITxPool _txPool;
+    private readonly IBlockProcessor _blockProcessor;
+    private readonly IBlockTree _blockTree;
+    private readonly IAccountStateProvider _accounts;
+    private readonly ILogger _logger;
+
+    private const int UnconfirmedBlocksCacheSize = 64;
+    private readonly LruCache<long, bool>
+        _unconfirmedBlocksCensorshipDetector = new(
+            UnconfirmedBlocksCacheSize, UnconfirmedBlocksCacheSize, "unconfirmedBlocksCensorshipDetector"
+            );
+    private const int CacheSize = 4;
+    private readonly LruCache<long, bool>
+        _censorshipDetector = new(CacheSize, CacheSize, "censorshipDetector");
+
+    public CensorshipDetector(
+        ITxPool txPool,
+        IBlockProcessor blockProcessor,
+        IBlockTree blockTree,
+        IAccountStateProvider accounts,
+        ILogManager logManager)
+    {
+        _txPool = txPool;
+        _blockProcessor = blockProcessor;
+        _blockTree = blockTree;
+        _accounts = accounts;
+        _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+        _blockProcessor.BlockProcessing += OnBlockProcessing;
+        _blockTree.BlockAddedToMain += OnBlockAdded;
+    }
+
+    private void OnBlockProcessing(object? sender, BlockEventArgs e)
+    {
+        Task.Run(() => PreCachingOperation(e.Block));
+    }
+
+    private void OnBlockAdded(object? sender, BlockEventArgs e)
+    {
+        Task.Run(() => Cache(e.Block));
+    }
+
+    private void PreCachingOperation(Block block)
+    {
+        UInt256 maxGasPriceInBlock = 0;
+        Transaction[] blockTransactions = block.Transactions;
+        foreach (Transaction tx in blockTransactions)
+        {
+            if (!tx.SupportsBlobs)
+            {
+                if (tx.Supports1559)
+                {
+                    maxGasPriceInBlock = UInt256.Max(maxGasPriceInBlock, tx.GasPrice);
+                }
+                else
+                {
+                    maxGasPriceInBlock = UInt256.Max(maxGasPriceInBlock, tx.GasPrice + block.BaseFeePerGas);
+                }
+            }
+        }
+
+        Transaction bestTx = _txPool.GetBestTxOfEachSender().First();
+        UInt256 maxGasPriceInPool = bestTx.Supports1559 ? bestTx.GasPrice + block.BaseFeePerGas : bestTx.GasPrice;
```

Comment:
This tx can potentially not be ready (has nonce gap). All required info already is in `gasBottleneck`.
`UInt256 maxGasPriceInPool = bestTx.GasBottleneck;` would be enough, because it is already `gasPremium + baseFee` for 1559+ and `gasPrice` for others. And `gasBottleneck` would be 0 if there is a nonce gap

Reply:
```suggestion
        UInt256 maxGasPriceInPool = 0;
        foreach (Transaction tx in _txPool.GetBestTxOfEachSender())
        {
            if (tx.GasBottleneck is not null || tx.GasBottleneck > 0)
            {
                maxGasPriceInPool = (UInt256)tx.GasBottleneck;
                break;
            }
            else
            {
                continue;
            }
        }
```
How about this? We check if the GasBottleneck is calculated and the transaction is ready. If so, we simply set maxGasPriceInPool and break, instead of iterating throughout all best transactions.

Reply:
Good idea, perfectly logical. We don't even need to iterate further if best tx doesn't have `tx.GasBottleneck > 0` - it will mean that we will not find any tx with positive gas bottleneck. So we can focus only on first tx from this collection.

Going further - maybe we can get from `SortedPool` only one, best tx? We would need to create e.g. `GetBest()` method, which will call `GetFirsts()` and return only first element. Then we can replace `GetBestTxOfEachSender()` with e.g. `GetBestTx()`, which will return `GetBest()` from SortedPool